src/libraries/System.Text.Json/ref/System.Text.Json.cs

```diff
@@ -959,6 +960,14 @@ public static partial class JsonMetadataServices
         public static System.Text.Json.Serialization.JsonConverter<T> GetEnumConverter<T>(System.Text.Json.JsonSerializerOptions options) where T : struct { throw null; }
         public static System.Text.Json.Serialization.JsonConverter<T?> GetNullableConverter<T>(System.Text.Json.Serialization.Metadata.JsonTypeInfo<T> underlyingTypeInfo) where T : struct { throw null; }
     }
+    public struct JsonParameterClrInfo
```

Comment:
Is there a reason this is a `struct`? It is mutable and has many properties. It feels more like this should be a `class`.

Reply:
Change the setters to init-only? I believe a `struct` is used to avoid an alloc.

Reply:
> used to avoid an alloc.

Is it worth it though? Aren't these only ever created at startup, cached forever, and never GC'd? Does a `struct` really help here?

Reply:
Yup the setters should be init-only. I think it's worth it to make this a struct (along with related proposed types in https://github.com/dotnet/runtime/issues/45448#issuecomment-887926294). Early start-up [measurements](https://gist.github.com/layomia/726d1071690df0d26e39807512f9cbee) showed a regression from the dynamic serializer to source-gen at one point where we were accidentally allocating more than the regular serializer would, which would be the case if we make these and related types classes.

Reply:
For this PR to get the ctor support in, I've changed the type to a class. In a subsequent PR to bring in the rest of the changes discussed in https://github.com/dotnet/runtime/issues/45448#issuecomment-892099426, I'll switch to a readonly or ref struct if perf numbers indicate that we should.

Reply:
I'll provide some perf numbers soon for class vs. struct vs. ref struct. Each POCO type and each property requires this metadata, so a POCO with 100 properties would require 101 temporary allocs. Ideally this are stack-based allocs for performance using `ref struct`, but we need to at least measure the impact on Blazor client to see if it is even measurable.

Reply:
The perf numbers running under Blazor client actually show a "PropertyInfo `class` is 2x faster than `struct` or `ref struct` even when passed byref (tested byval, "in", and "ref").

Ideally, to avoid the extra alloc, we expose the underlying `JsonPropertyInfo` class and just set the raw properties, but I don't think that is feasible given the v6 timeline.

Reply:
@steveharter thanks for running the tests. This PR makes the metadata info types classes, so we're currently in alignment with best perf numbers. Given this, I think exposing the underlying metadata classes (e.g. `JsonPropertyInfo`) is still valid for consideration and can help with future goals around users setting metadata themselves.